.github/workflows/selfhosted.yml

```diff
@@ -0,0 +1,122 @@
+# CI workflow that runs on self hosted cluster
+name: Self-hosted rules
+
+on:
+  workflow_call:
+    inputs:
+      used-labels:
+        description: 'Label names in JSON format'
+        required: true
+        type: string
+      python-version:
+        description: 'Used python version'
+        required: true
+        type: string
+      server-url:
+        description: ''
+        required: true
+        type: string
+      repository:
+        description: ''
+        required: true
+        type: string
+      run-id:
+        description: ''
+        required: true
+        type: string
+      poetry-extras:
+        description: 'poetry extra flags'
+        required: false
+        type: string
+        default: ""
+    secrets:
+      repo_token:
+        required: true
+
+jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - id: set-matrix
+        env:
+          LABELS: ${{ inputs.used-labels }}
+        run: |
+          platforms="{\"platform\":["
+          combined=""
+          shopt -s lastpipe
+          jq -c '.[]' <<< "$LABELS" | while read label; do
+              platform=(${label//-/ })
+              if [ ${platform[0]} == "\"run" ] && [ ${platform[1]} == "on" ]; then
+                platforms="${platforms}${combined:+,}\"${platform[2]}"
+                combined=${platforms}
+              fi
+          done
+          platforms+="]}"
+          echo ${platforms}
+          echo matrix="${platforms}" >> $GITHUB_OUTPUT
+  build:
+    needs: prepare
+    strategy:
+      matrix: ${{ fromJSON(needs.prepare.outputs.matrix) }}
+      fail-fast: false
+    runs-on: self-hosted
+    steps:
+    - name: Cleanup workspace manually
+      run: |
+        rm -rf _work/*
+    - uses: actions/checkout@v3
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ inputs.python-version }}
+    - name: Install and configure poetry
+      uses: snok/install-poetry@v1
+      with:
+        virtualenvs-create: false
+    - name: Load cached venv
+      id: cached-pip-wheels
+      uses: actions/cache@v3
+      with:
+        path: ~/.cache
+        key: venv-${{ runner.os }}-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/poetry.lock') }}
+    - name: Install package
+      run: |
+        python -m venv testenv
+        source testenv/bin/activate
+        poetry install --no-interaction ${{ inputs.poetry-extras }}
+    - name: Install simulation dependencies
+      if: ${{ matrix.platform == 'sim' }}
+      run: |
+        python -m pip install --upgrade pip
+        pip install cupy-cuda11x
+        pip install cuquantum-python-cu11
+        pip install git+https://github.com/qiboteam/qibojit
```

Comment:
I just have one question, since these dependencies are generally useful for running GPU tests, can we move them to `.toml` files?

If you want, you can update the docs adding this workflow.

Reply:
In principle, you can also add them as a Poetry group, or even an extra (Poetry groups are supposed to be used only in development, extras are available also to users of the deployed package).

Reply:
That's a good point, thanks. I tried adding
```toml
[tool.poetry.group.cuda11]
optional = true

[tool.poetry.group.cuda11.dependencies]
cupy-cuda11x = "^12.0.0"
cuquantum-python-cu11 = "^23.3.0"
qibojit = { git = "https://github.com/qiboteam/qibojit.git" }
```
in qibo's `pyproject.toml`, however when I do
```
poetry install --with cuda11
```
I get
```
Because qibo depends on cupy-cuda11x (^12.0.0) which doesn't match any versions, version solving failed.
```
even though this [version should exist in pypi](https://pypi.org/project/cupy-cuda11x/#history). @AleCandido do you have any idea how to fix this?

Reply:
Try to delete the cache https://python-poetry.org/docs/cli/#cache-clear

Reply:
This is fixed now and works in the workflow (cupy and cuquantum are installed properly): https://github.com/qiboteam/qibo/actions/runs/4962098439/jobs/8879734147?pr=894